index.js

```diff
@@ -1,149 +1,189 @@
+    if (opts.autoConnect) setImmediate(this.connect.bind(this));
+  }
+
+  connect() {
+    try {
+      this.connecting = true;
+      this.redis.connect(this.opts.host, this.opts.port, this.onConnect, this.onDisconnect);
+    } catch(e) {
+      this.reconnect();
+    }
+  }
+
+  processQueue() {
+    if (this.queue.length > 0){
+      this.queue.forEach((cmd) => this.redis.redisCmd(cmd.args, cmd.cb));
+      this.queue = [];
+    }
+  }
+
+  reconnect() {
+    const {opts} = this;
+    if (opts.tryToReconnect === false || this.reconnects >= opts.maxRetries) {
+      this.emit('error', new Error('Disconnected, exhausted retries.'));
+      this.end();
+      return;
+    }
+
+    this.reconnects++;
+    this.emit('reconnecting', this.reconnects);
+    if (this.reconnectTimeoutId) clearTimeout(this.reconnectTimeoutId);
+    this.reconnectTimeoutId = setTimeout(this.connect.bind(this), opts.reconnectTimeout);
+  }
+
+  selectDb(cb) {
+    const dbNum = this.opts.db;
+    if (dbNum > 0) {
+      this.redis.redisCmd(['SELECT', dbNum], (e) => {
+        if (e) {
+          this.emit('error', e);
+          this.reconnect();
+          return;
+        }
+        cb();
+      });
+    } else {
+      setImmediate(cb);
+    }
+  }
+
+  sendAuth(cb) {
+    const {auth} = this.opts;
+    if (auth) {
+      this.redis.redisCmd(['AUTH', auth], (e) => {
+        if(e) {
+          this.emit('error', new Error('Wrong password!'));
+          this.reconnect();
+          return;
+        }
+        cb();
+      });
+    } else {
+      setImmediate(cb);
+    }
+  }
+
+  _onConnect(e) {
+    if (e) {
+      this.emit('error', e);
+      this.reconnect();
+      return;
+    }
+    if (this.destroyed) {
+      // end() while we were connecting!
+      this.redis && this.redis.disconnect();
+      return;
+    }
 
+    this.ready = true;
+    this.connecting = false;
+    this.sendAuth(() => {
+      this.selectDb(() => {
+        this.processQueue();
+
+        if (!this.readyFirstTime) {
+          this.readyFirstTime = true;
+          this.emit('ready');
+        }
+        this.reconnects = 0;
+        this.emit('connect');
+      });
+    });
+  }
+
+  _onDisconnect(e) {
+    if (this.destroyed) return;
+    if (e) {
+      this.emit('error', e);
+    }
+    this.ready = false;
+    this.connecting = false;
+    this.emit('disconnect');
+    this.reconnect();
+  }
+
+  rawCall(args, cb) {
+    if(!args || !Array.isArray(args)) {
+      throw new Error('first argument to rawCall() must be an Array');
+    }
+
+    if (typeof cb === 'undefined') {
+      cb = (e) => {
+        if (e) this.emit('error', e);
+      };
+    }
+
+    // If not connected, push to queue
+    if (!this.ready) {
+      this.queue.push({args, cb});
+      return this;
+    }
+
+    // Send cmd
+    this.redis.redisCmd(args, cb);
+    return this;
+  }
+
+  rawCallAsync(args) {
+    return new Promise((resolve, reject) => {
```

Comment:
Promise will work very slow

Reply:
It's opt-in, and native promises as of Node 8 are quite fast - still slower than callbacks of course, but an educated user can make the choice. Wrapping it explicitly (since we know it only has one argument) is faster than using generic wrappers like `Bluebird.promisify()` or `util.promisify()`.